core/src/main/java/org/apache/iceberg/variants/VariantBuilderBase.java

```diff
@@ -0,0 +1,504 @@
+    int fieldOffsetSize = sizeOf(dataSize); // Number of bytes for each field offset
+    int headerSize =
+        1 + sizeBytes + numElements * fieldIdSize + (numElements + 1) * fieldOffsetSize;
+
+    // Shift existing data to make room for header
+    buffer.shift(objStartPos, headerSize);
+
+    buffer.insertByte(
+        VariantUtil.objectHeader(isLarge, fieldIdSize, fieldOffsetSize),
+        objStartPos); // Insert header byte
+    buffer.insertLittleEndianUnsigned(
+        numElements, sizeBytes, objStartPos + 1); // Insert number of elements
+
+    // Insert field IDs and offsets
+    int fieldIdStart = objStartPos + 1 + sizeBytes;
+    int fieldOffsetStart = fieldIdStart + numElements * fieldIdSize;
+    for (int i = 0; i < numElements; i++) {
+      buffer.insertLittleEndianUnsigned(
+          fields.get(i).id, fieldIdSize, fieldIdStart + i * fieldIdSize);
+      buffer.insertLittleEndianUnsigned(
+          fields.get(i).offset, fieldOffsetSize, fieldOffsetStart + i * fieldOffsetSize);
+    }
+
+    // Insert the offset to the end of the data
+    buffer.insertLittleEndianUnsigned(
+        dataSize, fieldOffsetSize, fieldOffsetStart + numElements * fieldOffsetSize);
+  }
+
+  /**
+   * Completes writing an array to the buffer. Array values are already written, and this method
+   * inserts header including the header byte, number of elements, and field offsets.
+   *
+   * @param arrStartPos The starting position of the array values in the buffer.
+   * @param offsets The offsets for each array value.
+   */
+  protected void endArray(int arrStartPos, List<Integer> offsets) {
+    int dataSize = buffer.pos - arrStartPos; // Total byte size of the array values
+    int numElements = offsets.size();
+
+    boolean isLarge = numElements > 0xFF; // Determine whether to use large format
+    int sizeBytes = isLarge ? 4 : 1; // Number of bytes for the array size
+    int fieldOffsetSize = sizeOf(dataSize); // Number of bytes of each field offset
+    int headerSize = 1 + sizeBytes + (numElements + 1) * fieldOffsetSize; // header size
+    int offsetStart = arrStartPos + 1 + sizeBytes; // Start position for offsets
+
+    // Shift existing data to make room for header
+    buffer.shift(arrStartPos, headerSize);
+
+    buffer.insertByte(
+        VariantUtil.arrayHeader(isLarge, fieldOffsetSize), arrStartPos); // Insert header byte
+    buffer.insertLittleEndianUnsigned(
+        numElements, sizeBytes, arrStartPos + 1); // Insert number of elements
+
+    // Insert field offsets
+    for (int i = 0; i < numElements; i++) {
+      buffer.insertLittleEndianUnsigned(
+          offsets.get(i), fieldOffsetSize, offsetStart + i * fieldOffsetSize);
+    }
+
+    // Insert the offset to the end of the data
+    buffer.insertLittleEndianUnsigned(
+        dataSize, fieldOffsetSize, offsetStart + numElements * fieldOffsetSize);
+  }
+
+  protected ByteBufferWrapper getBuffer() {
```

Comment:
Style: Iceberg does not use `get` in accessor names and discourages its use in general. If there's a more helpful verb then it is generally better to use it. Otherwise we usually omit `get`.